### ISSUE TITLE
Free Wolfram Engine QE/Solve/CEX/etc. Backend?

### ISSUE THREAD

**Issue by nrfulton** [LS-Lab/KeYmaeraX-release#56]
@aplatzer @smitsch : Good news: the new [free-for-personal-use Wolfram Engine](https://www.wolfram.com/engine/) works as a QE backend with the following settings on Mac OS X:

```
MATHEMATICA_LINK_NAME = /Applications/Wolfram\ Engine.app/Contents/MacOS/WolframKernel              
MATHEMATICA_JLINK_LIB_DIR = /Applications/Mathematica.app/Contents/SystemFiles/Links/JLink/SystemFiles/Libraries/MacOSX-x86-64
```

So, pretty close to allowing a Mathematica backend without requiring a paid Mathematica license. There are two road blocks:

1 (hard). As far as I could find, there is no `libJLinkNativeLibrary.jnilib` bundled with Wolfram Engine. Maybe I didn't search hard enough, or maybe this is something that our users can obtain separately without getting a Mathematica license?

2 (easy). Users cannot specify these settings in the web ui because the web ui expects something that ends in MathKernel, not WolframKernel.

**Comment by bbohrer** [comment]
@aplatzer I just downloaded and installed for Windows. The MathKernel and libJLinkNativeLibrary.dll exist there with their usual names and it "just works" after configuring it in the UI (well, and after logging in to your Wolfram account on the command line).

Has anybody checked it on other OS's?
@nrfulton @smitsch we're trying to figure this out before giving a demo to folks here. If it's this different across platforms though, maybe it's best to do a demo that doesn't need it :/
@nrfulton Did the machine also have a Mathematica install? And did you activate/run the wolfram shell? Just wondering if anything is different between our machines except OS hmm

**Comment by nrfulton** [comment]
I didn't see the jlink jnlp file on Mac OS, but I was busy with two deadlines so didn't look very hard. Maybe it's there but in a non standard location. I'll check again today.

I'll also look into Ubuntu tonight.

**Comment by nrfulton** [comment]
The machine did have mathematica and I did have to run/activate the she'll per Wolfram instructions.

**Comment by bbohrer** [comment]
There's no need for jnlp, I assume you mean the shared object file.
.dll on windows,.so on Linux, idk what it is on Mac

**Comment by nrfulton** [comment]
Yeah, I meant `libJLinkNativeLibrary.jnilib` on Mac... was walking and not able to look it up.

**Comment by nrfulton** [comment]
Fortunately, though, it is bundled with the engine! Here's where it is on Mac (note: this is different from normal):

`/Applications/Wolfram Engine.app//Contents/Resources/Wolfram Player.app/Contents/SystemFiles/Links/JLink/SystemFiles/Libraries/MacOSX-x86-64/libJLinkNativeLibrary.jnilib`

**Comment by nrfulton** [comment]
Okay, I finally had some time today to investigate this on MacOS and Ubuntu.

### Linux

On Ubuntu 180.04 with a fresh VM, the following settings just work:
 * Kernel default location: `/usr/local/Wolfram/WolframEngine/12.0/Executables/MathKernel`
 * Mathematica J/Link native library: `/usr/local/Wolfram/WolframEngine/12.0//SystemFiles/Links/JLink/SystemFiles/Libraries/Linux-x86-64/libJlinkNativeLibrary.so`

No changes are needed except adding these to the search paths for default configuration suggestions.

### Mac OS
On Mac OS, the story is more complicated:

1. You need to use `/Applications/Wolfram Engine.app/Contents/MacOS/WolframKernel` (*note*: the `MathKernel` in `Wolfram Player.app` doesn't seem to work with *any* `jnilib`. This means line 404 of `Request.scala` needs a small change.
2. I still haven't been able to get the Wolfram Engine to work without using my full Mathematica installation's `jnilib`. In particular, the one in `/Applications/Wolfram Engine.app/Contents/Resources/Wolfram Player.app/Contents/SystemFiles/Links/JLink/SystemFiles/Libraries/MacOSX-x86-64/libJLinkNativeLibrary.jnili` doesn't seem to work.

I didn't try installing Wolfram Engine without a pre-existing Mathematica installation on Mac OS... maybe someone else can try that and let us know if then useful `jnilib` files magically show up?

### Todos to support Wolfram Engine in KeYmaera X

1. Add instructions to the tool configuration page on how to install and activate Wolfram Engine.
2. Add `|| linkNameFile.getName == "WolframKernel"` to line 404 of `keymaerax-webui/.../hydra/Request.scala`.
3. Add wolfram script to the possible suggested install locations for each platofrm.
4. Somehow figure out the MacOS story...

**Comment by nrfulton** [comment]
Closed by 18fc51cdcb1d48bc350fc269f33fb12efcb7bedc and 7c26f274c8b468bd9907692972083cfb0f5d4e5f
